src/k2/tso/service/Service.cpp

```diff
@@ -143,12 +79,94 @@ seastar::future<> TSOService::_collectWorkerURLs() {
     });
 }
 
-seastar::future<> TSOService::_cpoRegister() {
-    // TODO implement CPO-based registration and ID assignment
-    _tsoId = seastar::this_shard_id() + 100000;
+seastar::future<> TSOService::_assign(uint64_t tsoID, k2::Duration errBound) {
+    _CPOErrorBound = errBound;
+    return seastar::make_ready_future<>()
+        .then([this]{
+            seastar::future<> startFut = seastar::make_ready_future<>();
+            if (seastar::this_shard_id() == 0) {

```

Comment:
I'm wondering if all this shard 0 logic should be part of the assignment logic. If the CPO happens to not assign shard 0 then none of the shards will be able to operate. We can ask Ivan about this too

Reply:
Yeah it make sense. Ivan talked with me about this and I think I'll update this to use a `test_and_set` condition on an atomic flag.